Short links

clbn opened the pull request:
This is the server-side part of the [Short Links](https://github.com/FreeFeed/freefeed-server/issues/542) feature.

- [x] Add options for the short links to the config file
- [x] Add the table to the DB
- [x] Add code for generating short IDs when creating posts
- [x] Add `shortId` prop to Post objects in API responses
- [x] Update single-post API endpoint (`GET /v2/posts/:postId`) to accept post's short ID along with the long ID
- [x] Tests!
- [x] Update backlinks extraction
- [x] Write a script for backfilling short IDs (for old posts)

Part 2:

- [x] Add support to EventService (notifications) 
- [x] Add support for realtime events
- [x] Add support to bin/reindex_backlinks

clbn committed:
short-links: Add DB structure and generating short IDs for new posts

davidmz commented on types/config.d.ts:
```diff
@@ -59,6 +59,12 @@ declare module 'config' {
+
+    postShortIds: {
+      initialLength: number;
+      stopWords: [string];
```
`[string]` → `string[]`

clbn replied:
Fixed.

davidmz commented on data_transfer/PgAdapter.js:
```diff
@@ -1,5 +1,6 @@
 import _ from 'lodash';
 import validator from 'validator';
+import config from 'config';
 
```
Проблема в том, что это не настоящий PgAdapter:) data_transfer — это временный каталог с кодом для переезда с редиса на постгрес, его давно пора совсем удалить. А настоящий адаптер лежит в каталоге /app/support/DbAdapter

clbn replied:
Fixed.

davidmz commented on data_transfer/PgAdapter.js:
```diff
@@ -341,10 +342,54 @@ export class PgAdapter {
+  isStringDecent = (str) => !config.postShortIds.stopWords.some((word) => str.includes(word));
+
+  getRandomString = (length) =>
+    [...Array(length)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');
```
Ой, не надо так. Лучше crypto.randomBytes

clbn replied:
Не думаю, что это необходимо, но не готов спорить. Исправил с `crypto`.

davidmz commented on app/support/DbAdapter/posts.js:
```diff
@@ -22,8 +25,9 @@ const postsTrait = (superClass) =>
-      const res = await this.database('posts').returning('uid').insert(preparedPayload);
-      return res[0];
+      const [longId] = await this.database('posts').returning('uid').insert(preparedPayload);
+      await this.createPostShortId(longId);
```
Слушай, а давай это всё-таки завернём в транзакцию? Иначе сбой между этими двумя строчками приведёт к появлению поста без shortId.

davidmz replied:
Причём не просто транзакцию, а с `lock table post_short_ids in share row exclusive mode`, чтобы точно в post_short_ids только один процесс писал.

clbn replied:
Сделал это отдельным коммитом, проверь, пожалуйста.

clbn committed:
short-links: Add optional shortId field to Post objects in API responses

davidmz commented on app/support/DbAdapter/posts.js:
```diff
@@ -22,7 +25,19 @@ const postsTrait = (superClass) =>
+        await trx.raw('lock table post_short_ids in share row exclusive mode');
+
+        // Create post
+        const [postId] = await trx('posts').insert(preparedPayload).returning('uid');
```
![image](https://user-images.githubusercontent.com/132120/150192455-65f60b4d-26bd-4a61-8420-63048a1de738.png)
Надо всё это писать с включённым eslint

clbn replied:
Ага, пришлось IDEA проапгрейдить, чтоб встроенный eslint заработал. Пушнул исправленное.

davidmz commented on app/support/DbAdapter/posts.js:
```diff
@@ -469,6 +529,7 @@ const POST_COLUMNS_MAPPING = {
 
 export const POST_FIELDS = {
   uid: 'id',
+  short_id: 'shortId',
```
Здесь это не имеет смысла т. к. в таблице posts нет поля short_id

clbn replied:
@davidmz А как тогда оно превратит `short_id` в `shortId`? Вроде же `friendfeed_url` полностью аналогичное поле?

davidmz replied:
А, да, ты прав… Там оно именно для этого.

davidmz commented on app/support/DbAdapter/timelines-posts.js:
```diff
@@ -322,6 +322,7 @@ const timelinesPostsTrait = (superClass) =>
         'comments_count',
         'likes_count',
         'friendfeed_url',
+        'short_id',
```
Соответственно, тут это тоже не нужно

clbn committed:
short-links: Wrap short_id creation into a transaction

clbn committed:
Simplify postAccessRequired (remove unnecessary parameter)

clbn committed:
short-links: Support short_id in API (route /v2/posts/:postId)

clbn committed:
short-links: Update migration timestamp

clbn committed:
short-links: Add integration tests

clbn committed:
short-links: Add support in backlinks

davidmz commented on app/support/backlinks.ts:
```diff
@@ -16,6 +16,11 @@ export function extractUUIDs(text: string | null): UUID[] {
   return text ? [...text.matchAll(uuidRe)].map((m) => m[0]).filter(onlyUnique) : [];
 }
 
+export const shortLinkRe = /\/[A-Za-z0-9]{3,25}\/([0-9a-f]{6,10})/gi;
```
Тут ещё надо минус учесть, для groupnames

davidmz replied:
И не надо ли границу слов как-то проверять? А то оно и в посторонних урлах может сработать. Или понадеемся, что 6 символов с нашими id-ами вряд ли совпадут?

clbn replied:
Тут непонятно. С одной стороны, мы вроде как хотели всю эту логику на клиент отдать. С другой стороны, из-за бэклинков вот приходится делать парсинг на сервере. Надо бы как-то формализовать это, чтоб алгоритм на сервере и клиенте — ну, не совпадали, но хоть как-то были близки.

Или не париться, потому что к большому беспорядку false positives не приведут…

davidmz commented on test/integration/support/DbAdapter/backlinks.js:
```diff
@@ -72,8 +72,8 @@ describe('Backlinks DB trait', () => {
-    before(() => venus.update({ isProtected: '1', isPrivate: '0' }));
-    after(() => venus.update({ isProtected: '0', isPrivate: '0' }));
+    before(async () => await venus.update({ isProtected: '1', isPrivate: '0' }));
+    after(async () => await venus.update({ isProtected: '0', isPrivate: '0' }));
```
Вообще, оно и так должно работать, before/after вполне понимают, когда им Promise возвращаешь

clbn replied:
Понятно, тогда уберу.

davidmz commented on test/integration/support/DbAdapter/backlinks.js:
```diff
@@ -246,9 +249,9 @@ describe('Backlinks DB trait', () => {
-    after(() => {
-      marsPost.update({ body: `luna post: example.com/${lunaPost.id}` });
-      jupiterCommentNo2.destroy();
+    after(async () => {
```
Вот только тут нужен явный async (ну или можно с Promise.all выпендриться:))

clbn committed:
Fix async test for backlinks

clbn committed:
short-links: Add script for backfilling short IDs

clbn committed:
short-links: Add support to bin/reindex_backlinks

clbn committed:
short-links: Add support to EventService (notifications)

clbn committed:
short-links: Add support for realtime events

davidmz commented on bin/reindex_backlinks.js:
```diff
@@ -124,6 +131,20 @@ process.stdout.write(`\n`);
 
+              // Only the real post Short Ids
+              const shortIdRecords = await dbAdapter.database.getAll(
+                `select long_id, short_id from post_short_ids where short_id = any(:shortIds)`,
```
Не надо ли тут добавить `and long_id is not null`?

clbn replied:
Точно, мы же удаляем посты, оставляя короткие. Сейчас в двух местах исправлю.

clbn replied:
@davidmz Исправил.

clbn committed:
short-links: Fix two SQL queries for backlinks

clbn committed:
short-links: Add comment links support to EventService (notifications)

This change adds parsing of short comment links (e.g. `/user/9bac13#ad2b`)
to EventService a.k.a Notifications. That is the only place we need
special support. Other relevant server parts (backlinks table, reindex
script, realtime events) only need the post_id anyway (for now), so they
work with such links already, not caring about the comment_id part.

clbn committed:
short-links: Define Comment.shortId, add it to API

We define short_comment_id as two characters of the UUID, followed by
the hex-encoded seq_number.

clbn committed:
short-links: Revamp Comment.shortId

Before:
- generate shortId on READ
- use 2 chars of UUID followed by seqNumber (resulting length is 3+)

After:
- generate shortId on WRITE, and put into table `comments` (column `short_id`)
- use randomly generated hex string (resulting length is 4-6)

clbn committed:
short-links: Update changelog
